Firestore/core/src/firebase/firestore/util/async_queue.h

```diff
@@ -0,0 +1,296 @@
+
+    if (HasDue()) {
+      DoPop(out, scheduled_.begin());
+      return true;
+    }
+    return false;
+  }
+
+  // Removes the first entry satisfying predicate from the queue, moves it into
+  // `out`, and returns true. If no such entry exists, doesn't modify `out` and
+  // returns false. Predicate is applied to entries in order according to their
+  // scheduled time. `out` may be `nullptr` (in which case the value will be
+  // simply discarded).
+  //
+  // Note that this function doesn't take into account whether the removed entry
+  // is past its due time.
+  template <typename Pred>
+  bool RemoveIf(T* const out, const Pred pred) {
+    std::lock_guard<std::mutex> lock{mutex_};
+    const auto found =
+        std::find_if(scheduled_.begin(), scheduled_.end(),
+                     [&pred](const Entry& s) { return pred(s.value); });
+    if (found != scheduled_.end()) {
+      DoPop(out, found);
+      return true;
+    }
+    return false;
+  }
+
+  // Blocks until at least one entry is available for which the scheduled time
+  // is due now (according to the system clock), removes the entry which is the
+  // most overdue from the queue and moves it into `out`. The function will
+  // attempt to minimize both the waiting time and busy waiting. `out` may be
+  // `nullptr` (in which case the value will be simply discarded).
+  void PopBlocking(T* const out) {
+    std::unique_lock<std::mutex> lock{mutex_};
+
+    while (true) {
+      cv_.wait(lock, [this] { return !scheduled_.empty(); });
+
+      // To minimize busy waiting, sleep until either the nearest entry in the
+      // future either changes, or else becomes due.
+      const auto until = scheduled_.front().due;
+      cv_.wait_until(lock, until,
+                     [this, until] { return scheduled_.front().due != until; });
+      // There are 3 possibilities why `wait_until` has returned:
+      // - `wait_until` has timed out, in which case the current time is at
+      //   least `until`, so there must be an overdue entry;
+      // - a new entry has been added which comes before `until`. It must be
+      //   either overdue (in which case `HasDue` will break the cycle), or else
+      //   `until` must be reevaluated (on the next iteration of the loop);
+      // - `until` entry has been removed. This means `until` has to be
+      //   reevaluated, similar to #2.
+
+      if (HasDue()) {
+        DoPop(out, scheduled_.begin());
+        return;
+      }
+    }
+  }
+
+  bool empty() const {
+    std::lock_guard<std::mutex> lock{mutex_};
+    return scheduled_.empty();
+  }
+
+  size_t size() const {
+    std::lock_guard<std::mutex> lock{mutex_};
+    return scheduled_.size();
+  }
+
+ private:
+  struct Entry {
+    bool operator<(const Entry& rhs) const {
+      return due < rhs.due;
+    }
+
+    T value;
+    TimePoint due;
+  };
+  // All removals are on the front, but most insertions are expected to be on
+  // the back.
```

Comment:
Alternatively, a `priority_queue` could be used, but it doesn't provide stable ordering of equivalent elements. I'd avoid cluttering this class with any tie-breaking "identifiers".